File: mixin-padang-webface/src/main/js/padang/functions/dataset/JoinRows.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import PlanUtils from "webface/plan/PlanUtils";

import FunctionPlan from "padang/plan/FunctionPlan";
import ParameterPlan from "padang/plan/ParameterPlan";
import ParameterPlanUtils from "padang/plan/ParameterPlanUtils";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import ColumnFunction from "padang/functions/dataset/ColumnFunction";
import DatasetFunction from "padang/functions/dataset/DatasetFunction";

export default class JoinRows extends DatasetFunction {

    public static FUNCTION_NAME = "JoinRows";

    public static TYPE_INNER = "inner";
    public static TYPE_LEFT = "left";
    public static TYPE_RIGHT = "right";
    public static TYPE_FULL = "outer";

    constructor(
        public dataset: any,
        public leftKeys: any[],
        public rightDataset: any,
        public rightKeys: any[],
        public type: string) {
        super(JoinRows.FUNCTION_NAME, dataset);
    }

    public static LEFT_KEYS_PLAN: ParameterPlan = ParameterPlanUtils.createListPlan(
        "leftKeys",
        "Left Keys",
        "Left Key list",
        ColumnFunction.SPECIFIED_PLAN
    );

    public static RIGHT_DATASET_PLAN: ParameterPlan = ParameterPlanUtils.createAnyPlan(
        "rightDataset",
        "Right Dataset",
        "Right Dataset",
        DatasetFunction.DATASET_NAMES_ASSIGNABLE
    );

    public static RIGHT_KEYS_PLAN: ParameterPlan = ParameterPlanUtils.createListPlan(
        "rightKeys",
        "Right Keys",
        "Right key list",
        PlanUtils.createSpecifiedPlan(
            "column",
            PlanUtils.createTextPlan()
                .setAssignable(ColumnFunction.getColumnKeysAssignable(JoinRows.RIGHT_DATASET_PLAN.getName()))
        )
    );

    public static TYPE: ParameterPlan = ParameterPlanUtils.createTextPlan(
        "type",
        "type",
        "Join Type",
        JoinRows.TYPE_INNER,
        "=[" +
        "'" + JoinRows.TYPE_INNER + "', " +
        "'" + JoinRows.TYPE_LEFT + "', " +
        "'" + JoinRows.TYPE_RIGHT + "', " +
        "'" + JoinRows.TYPE_FULL + "']"
    );

    public static getPlan(): FunctionPlan {
        let plan = new FunctionPlan(
            JoinRows.FUNCTION_NAME,
            "Join Rows",
            "mdi-set-none",
            "Join rows based on keys and foreign keys"
        );
        let parameters = plan.getParameterList();
        parameters.add(DatasetFunction.DATASET_PLAN);
        parameters.add(JoinRows.LEFT_KEYS_PLAN);
        parameters.add(JoinRows.RIGHT_DATASET_PLAN);
        parameters.add(JoinRows.RIGHT_KEYS_PLAN);
        parameters.add(JoinRows.TYPE);
        return plan;
    }

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(JoinRows.getPlan());

File: mixin-padang-garut/src/main/python/garut/functions/dataset/join_rows.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.#
from typing import Dict, List

from pandas import DataFrame
from garut.bearer import Bearer
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class JoinRows(DatasetFunction):

    FUNCTION_NAME = "JoinRows"

    TYPE_INNER = "inner"
    TYPE_LEFT = "left"
    TYPE_RIGHT = "right"
    TYPE_FULL = "outer"

    LEFT_KEYS = "leftKeys"
    RIGHT_DATASET = "rightDataset"
    RIGHT_KEYS = "rightKeys"
    TYPE = "type"

    def execute(self, context: Context, options: Dict[str, any]) -> any:

        # Datasets
        leftDataset = self.getDataset(options)
        rightDataset: DataFrame = options.get(JoinRows.RIGHT_DATASET, None)
        if isinstance(rightDataset, Bearer):
            rightDataset = rightDataset.getResult()

        # Keys
        leftKeys: List[str] = options[JoinRows.LEFT_KEYS]
        rightKeys: List[str] = options[JoinRows.RIGHT_KEYS]

        # Index
        type = options.get(JoinRows.TYPE, JoinRows.TYPE_INNER)
        suffixes = ["_left", "_right"]
        rightIndexed = rightDataset.set_index(rightKeys)

        # Merge
        result = leftDataset.merge(
            rightIndexed, how=type, left_on=leftKeys, right_on=rightKeys, right_index=True, suffixes=suffixes
        )

        # Type setting
        for i in range(len(leftKeys)):
            leftKey = leftKeys[i]
            rightKey = rightKeys[i]
            if leftDataset[leftKey].dtype == rightDataset[rightKey].dtype:
                result[leftKey] = result[leftKey].astype(leftDataset[leftKey].dtype)

        return result


parameters = [DatasetFunction.DATASET, JoinRows.LEFT_KEYS, JoinRows.RIGHT_DATASET, JoinRows.RIGHT_KEYS, JoinRows.TYPE]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(JoinRows.FUNCTION_NAME, JoinRows, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(JoinRows.FUNCTION_NAME)
